packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -0,0 +1,434 @@
+        if (typeof key === "symbol") {
+			switch (key) {
+				case proxySymbol:
+				case type:
+                case getTypeName:
+				// Currently not supporting iteration over fields.
+                // case Symbol.iterator:
+					return true;
+                case value:
+                    // Could do `target.value !== ValueSchema.Nothing`
+                    // instead if values which could be modified should report as existing.
+                    return target.value !== undefined;
+				default:
+					return false;
+			}
+		}
+
+        // For now primary array fields are handled by just returning the array, so we don't need this:
+        // const length = target.getPrimaryArrayLength();
+        // if (length !== undefined) {
+        //     // Act like an array.
+        //     // This means that "0" can be present, but not "0.0", "0.1", "-0", " 0" etc.
+        //     // Simplest way to check for this is to round trip:
+        //     if (key in []) {
+        //         return true;
+        //     }
+        //     const numeric = Number(key);
+        //     if (String(Number(key)) === key && Number.isInteger(numeric) && numeric >= 0 && numeric < length) {
+        //         return true;
+        //     }
+        // }
+
+		return target.has(key);
+	},
+    // Includes all non-empty fields, which are the enumerable fields.
+	ownKeys: (target: ProxyTarget): string[] => {
+        return target.keys;
+	},
+	getOwnPropertyDescriptor: (target: ProxyTarget, key: string | symbol): PropertyDescriptor | undefined => {
+		// We generally don't want to allow users of the proxy to reconfigure all the properties,
+        // but it is an TypeError to return non-configurable for properties that do not exist on target,
+        // so they must return true.
+
+        if (typeof key === "symbol") {
+			if (key === proxySymbol) {
+				return { configurable: true, enumerable: false, value: target, writable: false };
+			} else if (key === type) {
+				return { configurable: true, enumerable: false, value: target.type, writable: false };
+			} else if (key === getTypeName) {
+                return {
+                    configurable: true, enumerable: false, value: target.getTypeName.bind(target), writable: false,
+                };
+            }
+		} else if (target.has(key)) {
+            return {
+                configurable: true,
+                enumerable: true,
+                value: target.proxifyField(key),
+                writable: false,
+            };
+		}
+		return undefined;
+	},
+};
+
+/**
+ * See {@link UnwrappedEditableTree} for documentation on what unwrapping this perform.
+ */
+function inProxyOrUnwrap(target: ProxyTarget): UnwrappedEditableTree {
+    if (isPrimitive(target.type)) {
+        const nodeValue = target.cursor.value;
+        // see `isPrimitive` doc for `undefined` case
+        assert(nodeValue !== undefined, "Undefined values not allowed");
+        if (isPrimitiveValue(nodeValue)) {
```

Comment:
This assert is not valid. A node which has ValueSchema.Serializable can have an undefined value here, and we need to support that.

How that should be handled is isPrimitiveValue should return false when given undefined (or should have its type signature adjusted to now allow passing in undefined).

Looks like the current implementation of isPrimitiveValue has a bug where it returns true when given undefined: fix that, and remove this assert and it should work properly.

```suggestion
        if (isPrimitiveValue(nodeValue)) {
```

Reply:
To be honest, I thought that `Serializable` was omitted on purpose for this primitive node
https://github.com/microsoft/FluidFramework/pull/11589/files?diff=unified&w=1#diff-ba26d1ea70456ae7a727078e93a12eb306cd25117730f41aa9aeca3093f80c32R71-R74

The problem is imho if one will accidentally provide `undefined` instead of a primitive value, and this will be somewhere deeply in the tree, it will be very troublesome to figure out what happens, as without any assertion (not necessarily this one) this code will silently return a proxy.

I've reworked this to assert if `undefined` node is `Serializable` (and fixed `isPrimitiveValue` as suggested below), but I don't like to allow undefined numbers or strings as this potentially could lead to an undefined behaviour.

Reply:
There are two definitions of "primitive" here. One at the schema level which just has a placeholder implementation (isPrimitive) which will eventually be replaced by metadata on the view schema (I intentionally made this overly general, so EditiableTree needs to handle cases it can't unwrap so that future changes to how this works won't break editiableTree), and a more strict definition at the node level that is logically part of the EditiableTree implementation which explicitly excludes undefined and objects from being treated as primitives.

Note that if you use ValueSchema's Number, Boolean or String, undefined is not allowed (would be out of schema), and thus you will unconditionally pass both checks and end up fully unwrapped.

This proxy can assume the data is in schema, and can (due does not have to) assert on out of schema cases. This assert could however be hit by data that is in schema and uses Serializable which is the problem.

The edge case here is that if the schema explicitly says Serializable, EditableTree need to handle all the allowed values for that (actual primitives, undefined, and objects). Since the tests currently arn't using Serializable, I decided it was fine for these cases to unwrap the actual primitive values, but keep the other cases (undefined and objects) in proxies. We can change this policy as long as we actually still handle all in schema cases without asserting.

I don't see any undefined behavior here: I think the behavior for this is defined in EditiableTreeOrPrimative's documentation.

Reply:
Em, I don't think you got my last point. With

> I've reworked this...

I mean, I have changed the code, so that the assert you mentioned initially is not there anymore, and the rest is exactly about to let `undefined` go throw if `Serializable` is explicitly defined in a schema, not letting primitive nodes with primitive values to be unwrapped accidentally just because someone put `undefined` into their value.

Reply:
I was not objecting to your change, just clarifying why I set it up how I did. Sorry for the confusion.